### ISSUE TITLE
Allow parallel GHC installs

### ISSUE THREAD

**Issue by ag-eitilt** [gentoo-haskell/gentoo-haskell#939]
Now that Cabal is pushing the new-style builds, it's (comparatively) easy to maintain multiple versions of GHC on a system at once -- whether for ensuring the package you're developing works on a good range of `base`, for keeping a more stable version around for system tasks while using the most recent for outgoing, or for building something that's not been updated for a while (different issue, but `uhc` can't actually build on anything >= 8.0; its ebuild needs to be updated to reflect that).

**Comment by trofi** [comment]
There are multiple facets to multislotted ghc:
1. installing multislot ghc (easy, we can make a separate ebuild for it as a symlink to unslotted ghc and have a few conditionals)
2. have all the library/binary dependencies right (quite hard, that's why we don't have slotted libraries either)
3. have haskell-updater detect breakages when any of ghc slots upgrades (moderate)

We can tackle [1.] first.

**Comment by ag-eitilt** [comment]
Idle thought for 2.: How possible is it to have dynamically-slotted ebuilds?  If we can make the libraries take their primary slot from the GHC version and their subslot (as they do now) from their own, would that work for dependency resolution?  Given the current install tree, the only file conflicts would be in `/usr/share/doc/`, and that can be fixed.  I know it's not currently implemented, but anyone who's worked on those tools, how tricky would it be to extend them (including convincing people it's a good idea, assuming it is one)?

**Comment by trofi** [comment]
It's not clear to me what "dynamic dependencies" are.

**Comment by pbarill** [comment]
It sure would be an interesting feature to get parallel GHC just like you can get this _easily_ on Gentoo with GCC and python. That way you could easily move on to the latest release while keeping an older version for legacy stuff. First, GCC releases are a lot more interchangeable than GHC. As for python, I'm no expert on the matter, but I see that concurrent versions work fine because each and every python package has a set of flags `python_target` to specify which version of python would you like this to be built with, when supported. People on that project are regularly (constantly?) dropping one old python version and adding a new one, based on newest stable and latest to be EOLed. That massive ongoing chore does not look interesting: "Hey it works with 3.6, 3.8 but I can't get it to work with 3.7 even though it says it should work ok. Help?"

The matter with GHC is that there's not much of an EOL concept (that I'm aware of). There's a release, and that's it, enjoy. (Just don't complain if you are stuck in the past.) So we would have to support something like vast sets of flags `ghc_target_80` all the way to ... `ghc_target_810`? I doubt that there is sufficient manpower to handle such a task. And whenever you add or modify a package, build/test it (build needed dependencies, rebuild all reverse dependencies), then repeat everything for each and every candidate GHC target. Yikes. I don't have access to a machine powerful enough to deal so many scenarios in a timely manner.

**Comment by ag-eitilt** [comment]
Yeah, figured it wouldn't be that simple.  The thought behind "dynamic dependencies" would be to generate multiple slots from a single ebuild via (or otherwise attached to) its dependency on GHC, so then as long as people set their `base` bounds correctly in the `*.cabal` we get the `ghc_target_*` testing for free.  I guess the first problem with that is that people don't set their bounds correctly.

But, yeah, anything as intensive as how Python is handled will need to wait until there's nearly as many people using Haskell, for us to pull maintainers from.

I will say that multiple GHC is working on my system, just so long as I only call the non-default versions through Cabal's sandboxing.  It's far from general-user friendliness, but it does work really nicely for dev work.

**Comment by necrose99** [comment]
eselect ghc ? works for me.

**Comment by khumba** [comment]
GHC and Cabal are packaged this way for Debian, maybe we can look at how these builds are configured for some inspriation:

https://downloads.haskell.org/~debian/

Having multiple GHCs available would be useful for development even if it's only for v2 or sandboxed builds to start with.  Thanks for posting your changes, @ag-eitilt, I'll be trying those out.

**Comment by necrose99** [comment]
**ghc-bin** make install ie **/usr/local.....** will have base binaries , and build remaining 
ghc-9.0.1-x86_64-deb10-linux-dwarf.tar.xz  , ghc-9.0.1-armv7-deb9-linux.tar.xz  ghc-9.0.1-aarch64-deb9-linux.tar.xz  

app-text/pandoc-bin  : https://gitlab.com/src_prepare/src_prepare-overlay/-/tree/master/app-text/pandoc-bin 
cabal-install-bin : https://downloads.haskell.org/cabal/cabal-install-3.4.0.0/ 


emerge ghc for arm64/etc  would then build regular ghc etc with option to clean ghc-bin / cabal-bin 


this would make instaling ghc on gentoo bit more easy , and or bootstrapping ghc...
